### PR TITLE
clickhouse-client: disallow usage of `--query` and `--queries-file` at the same time

### PR DESCRIPTION
**Pull request by Demilivor**
Closes #49769

And just as an idea:
Recent PR #49970 hints a better solution which is running from any "source" in the order of console arguments, e.g.:

```
# file.sql contains SELECT "File 2";
clickhouse-client --query "select 1" --queries-file "file.sql" --query "SELECT 3"

# Output
1
"File 2"
3
```

_This PR does not add this behavior, it only indicates a further improvement_

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
clickhouse-client now forbids usage of "--query" and "--queries-file" at the same time

### CODE REVIEW

**Comment by rvasin** [comment]
@rschu1ze Please be reviewer of this PR and assign 'can be tested' label.

**Comment by robot-ch-test-poll2** [comment]
This is an automated comment for commit 04560fe3ecc75d8451bc2a5b3772efab79c7fa18 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50210/04560fe3ecc75d8451bc2a5b3772efab79c7fa18/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: tests/queries/0_stateless/02751_multiquery_with_argument.sh [hunk 1]
```diff
@@ -13,6 +13,10 @@ $CLICKHOUSE_LOCAL --multiquery "SELECT 200; S" 2>&1 | grep -o 'Syntax error'
 $CLICKHOUSE_LOCAL --multiquery "; SELECT 201;" 2>&1 | grep -o 'Empty query'
 $CLICKHOUSE_LOCAL --multiquery "; S; SELECT 202" 2>&1 | grep -o 'Empty query'
 
+# Simultaneous --queries-file + --query (multiquery) is prohibited.
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02751_multiquery_with_argument.sh:16]: Cosmetic: let's add `passing` between `Simultaneously` and `--queries-file`.

File: tests/queries/0_stateless/01523_client_local_queries_file_parameter.sh [hunk 2]
```diff
@@ -21,4 +21,9 @@ DROP TABLE 01523_test;" > 01523_client_local_queries_file_parameter_tmp.sql
 
 echo -e "1,2\n3,4" | $CLICKHOUSE_LOCAL --queries-file=01523_client_local_queries_file_parameter_tmp.sql 2>&1
 
+# Simultaneous --queries-file + --query is prohibited.
```
Comment (rschu1ze) [inline tests/queries/0_stateless/01523_client_local_queries_file_parameter.sh:24]: Cosmetic: let's add `passing` between `Simultaneously` and `--queries-file`.

File: programs/client/Client.cpp [hunk 3]
```diff
@@ -1188,6 +1188,9 @@ void Client::processConfig()
     /// - --queries-file command line option is present.
     ///   The value of the option is used as file with query (or of multiple queries) to execute.
 
+    if (config().has("query") && config().has("queries-file"))
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Specify either `query` or `queries-file` option");
```
Comment (rschu1ze) [inline programs/client/Client.cpp:1192]: `Options '--query' and '--queries-file' cannot be specified at the same time`.

(same in LocalServer.cpp)

Comment (rschu1ze) [reply programs/client/Client.cpp:1192]: (I first thought that we should perhaps let Boost program_options do the check (similar to [this](https://stackoverflow.com/a/21914921)) but the disadvantage of that would be that we need to check against the short form `-q` as well. That isn't the case here as the options are translated to configuration already. But I would appreciate if you could add tests for the short form.)

File: programs/client/Client.cpp [hunk 4]
```diff
@@ -1188,6 +1188,9 @@ void Client::processConfig()
     /// - --queries-file command line option is present.
     ///   The value of the option is used as file with query (or of multiple queries) to execute.
 
+    if (config().has("query") && config().has("queries-file"))
```
Comment (rschu1ze) [inline programs/client/Client.cpp:1191]: I think the comment in l. 1183ff applies to l. 1194. Could you move l. 1191/1192 before the comment?

File: programs/client/Client.cpp (hunk 3 above)
Comment (Demilivor) [reply programs/client/Client.cpp:1192]: No need to check against short form of option in boost variables map (boost::program_options::variables_map). Arguments are already translated to long form after parsing.

Short arguments require check between parsing and storing into variables map, or even before boost options parsing.

Part of code from BaseClient.cpp
```cpp
    auto parser = po::command_line_parser(arguments)
                      .options(options_description.main_description.value())
                      .extra_parser(OptionsAliasParser(options_description.main_description.value()))
                      .allow_unregistered();
    po::parsed_options parsed = parser.run();
 
   // ...
   // po::parsed_options parsed can include both long and short forms of option
   // ...
   
    po::store(parsed, options);
   
   // after po::store options variables map will contain long form arguments
```

I will make test work both with short and long --query option variants

File: programs/client/Client.cpp (hunk 4 above)
Comment (Demilivor) [reply programs/client/Client.cpp]: Done.

File: programs/client/Client.cpp (hunk 3 above)
Comment (Demilivor) [reply programs/client/Client.cpp]: Done.

File: tests/queries/0_stateless/01523_client_local_queries_file_parameter.sh (hunk 2 above)
Comment (Demilivor) [reply tests/queries/0_stateless/01523_client_local_queries_file_parameter.sh]: Done.

File: tests/queries/0_stateless/02751_multiquery_with_argument.sh (hunk 1 above)
Comment (Demilivor) [reply tests/queries/0_stateless/02751_multiquery_with_argument.sh]: Done.

**Comment by rschu1ze** [comment]
01523_client_local_queries_file_parameter is [flaky](https://s3.amazonaws.com/clickhouse-test-reports/50210/2461d2ef0eb3a1fb1c09ef7be905702ab7e75371/stateless_tests_flaky_check__asan_.html). I think it is because it cannot safely run in parallel. Try to concatenate `01523_client_local_queries_file_parameter_tmp.sql` with a per-test-run random constant.

**Comment by Demilivor** [comment]
> 01523_client_local_queries_file_parameter is [flaky](https://s3.amazonaws.com/clickhouse-test-reports/50210/2461d2ef0eb3a1fb1c09ef7be905702ab7e75371/stateless_tests_flaky_check__asan_.html). I think it is because it cannot safely run in parallel. Try to concatenate `01523_client_local_queries_file_parameter_tmp.sql` with a per-test-run random constant.

I will check this this change next week. At least it looks error message relates to files

<details>
  <summary>Error message from logs</summary>
```
2023-05-26 04:24:25 rm: cannot remove '01523_client_local_queries_file_parameter_tmp.sql': No such file or directory
2023-05-26 04:24:25 , result:
2023-05-26 04:24:25 
2023-05-26 04:24:25 1
2023-05-26 04:24:25 2
2023-05-26 04:24:25 3
2023-05-26 04:24:25 [localhost] 2023.05.26 04:24:22.901921 [ 661 ] {500cac7d-35ae-4198-a0cb-270450ee3294} <Error> executeQuery: Code: 291. DB::Exception: Using file descriptor as source of storage isn't allowed for server daemons. (DATABASE_ACCESS_DENIED) (version 23.5.1.1) (from [::1]:37820) (comment: 01523_client_local_queries_file_parameter.sh) (in query: CREATE TABLE 01523_test (a Int64, b Int64) ENGINE = File(CSV, stdin);), Stack trace (when copying this message, always include the lines below):
2023-05-26 04:24:25 
2023-05-26 04:24:25 0. ./build_docker/./contrib/llvm-project/libcxx/include/exception:134: Poco::Exception::Exception(String const&, int) @ 0x0000000044282bb3 in /usr/bin/clickhouse
2023-05-26 04:24:25 1. ./build_docker/./src/Common/Exception.cpp:92: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x0000000024654b70 in /usr/bin/clickhouse
2023-05-26 04:24:25 2. ./build_docker/./contrib/llvm-project/libcxx/include/string:1499: DB::Exception::Exception<char const (&) [76], void>(int, char const (&) [76]) @ 0x00000000247192ab in /usr/bin/clickhouse
2023-05-26 04:24:25 3. ./build_docker/./src/Storages/StorageFile.cpp:0: DB::StorageFile::StorageFile(int, DB::StorageFile::CommonArguments) @ 0x0000000039adcfa5 in /usr/bin/clickhouse
2023-05-26 04:24:25 4. ./build_docker/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:35: std::shared_ptr<DB::StorageFile> std::allocate_shared[abi:v15000]<DB::StorageFile, std::allocator<DB::StorageFile>, int&, DB::StorageFile::CommonArguments&, void>(std::allocator<DB::StorageFile> const&, int&, DB::StorageFile::CommonArguments&) @ 0x0000000039b17421 in /usr/bin/clickhouse
2023-05-26 04:24:25 5. ./build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:962: std::shared_ptr<DB::IStorage> std::__function::__policy_invoker<std::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>::__call_impl<std::__function::__default_alloc_func<DB::registerStorageFile(DB::StorageFactory&)::$_0, std::shared_ptr<DB::IStorage> (DB::StorageFactory::Arguments const&)>>(std::__function::__policy_storage const*, DB::StorageFactory::Arguments const&) @ 0x0000000039afc282 in /usr/bin/clickhouse
2023-05-26 04:24:25 6. ./build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:0: DB::StorageFactory::get(DB::ASTCreateQuery const&, String const&, std::shared_ptr<DB::Context>, std::shared_ptr<DB::Context>, DB::ColumnsDescription const&, DB::ConstraintsDescription const&, bool) const @ 0x0000000039ad16b6 in /usr/bin/clickhouse
2023-05-26 04:24:25 7. ./build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:1417: DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::unique_ptr<DB::DDLGuard, std::default_delete<DB::DDLGuard>>&) @ 0x0000000037d79c82 in /usr/bin/clickhouse
2023-05-26 04:24:25 8. ./build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:0: DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) @ 0x0000000037d67150 in /usr/bin/clickhouse
2023-05-26 04:24:25 9. ./build_docker/./src/Interpreters/InterpreterCreateQuery.cpp:0: DB::InterpreterCreateQuery::execute() @ 0x0000000037d84833 in /usr/bin/clickhouse
2023-05-26 04:24:25 10. ./build_docker/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x0000000038df8204 in /usr/bin/clickhouse
2023-05-26 04:24:25 11. ./build_docker/./src/Interpreters/executeQuery.cpp:1176: DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x0000000038de96d7 in /usr/bin/clickhouse
2023-05-26 04:24:25 12. ./build_docker/./src/Server/TCPHandler.cpp:0: DB::TCPHandler::runImpl() @ 0x000000003b950134 in /usr/bin/clickhouse
```
</details>

**Comment by Demilivor** [comment]
> 01523_client_local_queries_file_parameter is [flaky](https://s3.amazonaws.com/clickhouse-test-reports/50210/2461d2ef0eb3a1fb1c09ef7be905702ab7e75371/stateless_tests_flaky_check__asan_.html). I think it is because it cannot safely run in parallel. Try to concatenate `01523_client_local_queries_file_parameter_tmp.sql` with a per-test-run random constant.

Done.

**Comment by rschu1ze** [comment]
Lgtm. CI won't let me merge, I restarted some tests. If that also doesn't help, please merge from `master` once more (that's also something the UI won't let me do on this PR for some reason).

**Comment by rschu1ze** [comment]
Silly me, I somehow added prefix "ClickHouse" to the restarted tests, with the result that 1. they are not restarted (because no such tests exist) and 2. they block merge.

@Demilivor Better click "Update branch", otherwise we'll be stuck for all eternity 😄

**Comment by rvasin** [comment]
> @Demilivor Better click "Update branch", otherwise we'll be stuck for all eternity 😄

Robert, it's not a problem @Demilivor  will update the branch in few hours, he is in GMT+10 time zone.

**Comment by Demilivor** [comment]
It looks this test stuck.
test_stuck

**Comment by rschu1ze** [comment]
CI is currently trash. Please don't merge master again, I restarted the failing tests manually.

**Comment by Demilivor** [comment]
> CI is currently trash. Please don't merge master again, I restarted the failing tests manually.

The reason for the falling tests is probably because of https://github.com/actions/runner/issues/2582

**Comment by rschu1ze** [comment]
ClickHouse Stateless Tests (tsan, S3 Storage) [3/5]: 02481_async_insert_race_long is [flaky](https://play.clickhouse.com/play?user=play#c2VsZWN0IHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMgYXMgc3RhdHVzLCByZXBvcnRfdXJsCmZyb20gY2hlY2tzCndoZXJlIHRvRGF0ZShjaGVja19zdGFydF90aW1lKSA+ICcyMDIzLTA1LTAyJyBhbmQgdGVzdF9uYW1lIGxpa2UgJyUwMjQ4MV9hc3luY19pbnNlcnRfcmFjZV9sb25nJScgYW5kIHN0YXR1cyA9PSAnRkFJTCcKb3JkZXIgYnkgY2hlY2tfc3RhcnRfdGltZSBkZXNj), #47866

**Pull request merged**